src/dragonball/src/device_manager/vfio_dev_mgr/mod.rs

```diff
@@ -0,0 +1,805 @@
+
+    fn id(&self) -> &str {
+        &self.hostdev_id
+    }
+
+    fn check_conflicts(&self, other: &Self) -> Result<()> {
+        if self.hostdev_id == other.hostdev_id {
+            return Err(VfioDeviceError::DeviceIDAlreadyExist(
+                self.hostdev_id.clone(),
+            ));
+        }
+
+        if !self.sysfs_path.is_empty() && self.sysfs_path == other.sysfs_path {
+            return Err(VfioDeviceError::DeviceAlreadyInUse(self.sysfs_path.clone()));
+        }
+
+        if !self.dev_config.bus_slot_func.is_empty()
+            && self.dev_config.bus_slot_func == other.dev_config.bus_slot_func
+        {
+            return Err(VfioDeviceError::DeviceAlreadyInUse(
+                self.dev_config.bus_slot_func.clone(),
+            ));
+        }
+
+        Ok(())
+    }
+}
+
+/// Vfio device info
+pub type VfioDeviceInfo = DeviceConfigInfo<HostDeviceConfig>;
+
+/// A device manager to manage all VFIO devices.
+pub struct VfioDeviceMgr {
+    vm_fd: Arc<VmFd>,
+    info_list: DeviceConfigInfos<HostDeviceConfig>,
+    locked_vm_size: u64,
+    vfio_container: Option<Arc<VfioContainer>>,
+    pci_vfio_manager: Option<Arc<PciSystemManager>>,
+    pci_legacy_irqs: Option<HashMap<u8, u8>>,
+    nvidia_shared_irq: Option<u32>,
+    logger: slog::Logger,
+}
+
+impl VfioDeviceMgr {
+    /// Create a new VFIO device manager.
+    pub fn new(vm_fd: Arc<VmFd>, logger: &slog::Logger) -> Self {
+        VfioDeviceMgr {
+            vm_fd,
+            info_list: DeviceConfigInfos::new(),
+            locked_vm_size: 0,
+            vfio_container: None,
+            pci_vfio_manager: None,
+            pci_legacy_irqs: Some(HashMap::new()),
+            nvidia_shared_irq: None,
+            logger: logger.new(slog::o!()),
+        }
+    }
+
+    /// Insert or update a VFIO device into the manager.
+    pub fn insert_device(
+        &mut self,
+        ctx: &mut DeviceOpContext,
+        config: HostDeviceConfig,
+    ) -> std::result::Result<(), VfioDeviceError> {
+        if !cfg!(feature = "hotplug") && ctx.is_hotplug {
+            return Err(VfioDeviceError::UpdateNotAllowedPostBoot);
+        }
+        slog::info!(
+            ctx.logger(),
+            "add VFIO device configuration";
+            "subsystem" => "vfio_dev_mgr",
+            "hostdev_id" => &config.hostdev_id,
+            "bdf" => &config.dev_config.bus_slot_func,
+        );
+        let device_index = self.info_list.insert_or_update(&config)?;
+        // Handle device hotplug case
+        if ctx.is_hotplug {
+            slog::info!(
+                ctx.logger(),
+                "attach VFIO device";
+                "subsystem" => "vfio_dev_mgr",
+                "hostdev_id" => &config.hostdev_id,
+                "bdf" => &config.dev_config.bus_slot_func,
+            );
+            let res = self.add_device(ctx, &config, device_index);
+            if let Err(e) = res {
+                return Err(e);
+            }
```

Comment:
Merge these into one line?

```suggestion
            self.add_device(ctx, &config, device_index)?;
```

Reply:
good catch!
fixed.